iris-mpc-gpu/src/server/actor.rs

```diff
@@ -844,16 +847,48 @@ impl ServerActor {
         ///////////////////////////////////////////////////////////////////
         // MERGE LEFT & RIGHT results
         ///////////////////////////////////////////////////////////////////
+
         tracing::info!("Joining both sides");
         // Merge results and fetch matching indices
         // Format: host_results[device_index][query_index]
-        self.distance_comparator.join_db_matches(
-            &self.db_match_list_left,
-            &self.db_match_list_right,
-            &self.final_results,
-            &self.current_db_sizes,
-            &self.streams[0],
-        );
+
+        // Initialize bitmap with OR rule, if exists
+        if !batch.or_rule_serial_ids.is_empty()
+            && !batch
+                .or_rule_serial_ids
+                .iter()
+                .all(|inner| inner.is_empty())
+        {
+            assert_eq!(batch.or_rule_serial_ids.len(), batch_size);
+
+            // Populate the pre-allocated OR policy bitmap with the serial ids
+            let host_or_policy_bitmap = prepare_or_policy_bitmap(
+                self.max_db_size,
+                batch.or_rule_serial_ids.clone(),
+                batch_size,
+            );
+
+            let device_or_policy_bitmap =
+                self.allocate_or_policy_bitmap(host_or_policy_bitmap.clone());
```

Comment:
let's maybe add some timing around this, so we now if this is worth improving more? (we had some ideas to improve this)